### PR TITLE
ng: make Polymer pagination configurable from Ng

### PR DESCRIPTION
**Pull request by stephanwlee**
This implements the pagination setting for Polymer based TensorBoard.
When changing the pagination setting, you can now change how many pages
of categories will be rendered on the Polymer side. Do note that
Angular based components will simply use selector to read the state from
the store instead.

TODO after the PR:
- stylize the settings dialog.

### CODE REVIEW

**Comment by stephanwlee** [comment]
Do note that this is a duplicate of #2813. Because of folder restructure refactors, the change became nearly un-salvageable and history got quite messy.

File: tensorboard/webapp/settings/polymer_interop_test.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (psybuzz) [inline tensorboard/webapp/settings/polymer_interop_test.ts]: nit: 2020

File: tensorboard/webapp/settings/polymer_interop_test.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  let store: MockStore<State>;
+  let setLimitCalls: number[];
+
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      providers: [provideMockStore()],
+      declarations: [SettingsPolymerInteropContainer],
+    }).compileComponents();
+
+    store = TestBed.get(Store);
```
Comment (psybuzz) [inline tensorboard/webapp/settings/polymer_interop_test.ts]: Prefer `TestBed.inject` for future Angular 9 compat?

File: tensorboard/webapp/settings/dialog_component.ts [hunk 3]
```diff
@@ -59,18 +81,45 @@ const getReloadPeriodInSec = createSelector(
...
+          type="number"
+          placeholder="Pagination Limit"
+          [formControl]="paginationControl"
+        />
+      </mat-form-field>
+      <mat-error
+        *ngIf="
+          paginationControl.hasError('min') ||
+          paginationControl.hasError('required') ||
+          paginationControl.hasError('integer')
```
Comment (psybuzz) [inline tensorboard/webapp/settings/dialog_component.ts]: Does `paginationControl.invalid` suffice?

File: tensorboard/webapp/reducer_config.ts [hunk 4]
```diff
@@ -36,8 +36,7 @@ export function logger(reducer: ActionReducer<any>): ActionReducer<any> {
   };
 }
 
-export const metaReducers: MetaReducer<any>[] =
-  config.env === 'dev' ? [logger] : [];
```
Comment (psybuzz) [inline tensorboard/webapp/reducer_config.ts]: Seems unrelated to this PR, is making the logger enabled in prod a deliberate change?

File: tensorboard/webapp/settings/polymer_interop_container.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  ngOnInit() {
+    this.getPageSize$
+      .pipe(
+        takeUntil(this.ngUnsubscribe),
+        distinctUntilChanged()
+      )
+      .subscribe((pageSize) => {
+        this.paginatedViewStore.setLimit(pageSize);
+      });
+  }
```
Comment (psybuzz) [inline tensorboard/webapp/settings/polymer_interop_container.ts]: When would the `settings-polymer-interop` ever be destroyed?  Its lifetime should be the same as the entire app.  Can we drop ngOnInit/Destroy?

File: tensorboard/webapp/app_container.ng.html [hunk 6]
```diff
@@ -19,3 +19,4 @@
 <plugins class="plugins"></plugins>
 <reloader></reloader>
 <hash-storage></hash-storage>
+<settings-polymer-interop></settings-polymer-interop>
```
Comment (psybuzz) [inline tensorboard/webapp/app_container.ng.html]: Is there some scenario when TensorBoard should not have `settings-polymer-interop`?  My understanding was that elements made available in this HTML template allow for customization, e.g. custom header.

If hash-storage and settings-polymer-interop are actually required functionality, it would be nice not to have to update each variant.  Wdyt about making some `<core-features></core-features>` that internally appends required fields hash-storage (which already lives in core/), settings-polymer-interop?  (not necessarily in this CL)

File: tensorboard/webapp/settings/polymer_interop_container.ts (hunk 5 above)
Comment (stephanwlee) [reply tensorboard/webapp/settings/polymer_interop_container.ts]: It depends on the user. As a component, you should not make decisions and omit cleanup based on users.

File: tensorboard/webapp/app_container.ng.html (hunk 6 above)
Comment (stephanwlee) [reply tensorboard/webapp/app_container.ng.html]: I don't feel too strongly about it but I am inclined _not to_ create `core-features`. If there is one component that you do not want, you will end up needing to fork it and copy-pasta anyways.

File: tensorboard/webapp/reducer_config.ts (hunk 4 above)
Comment (stephanwlee) [reply tensorboard/webapp/reducer_config.ts]: Oops! Reverted!

File: tensorboard/webapp/settings/dialog_component.ts (hunk 3 above)
Comment (stephanwlee) [reply tensorboard/webapp/settings/dialog_component.ts]: Yes, it does! Used.

File: tensorboard/webapp/settings/polymer_interop_container.ts (hunk 5 above)
Comment (psybuzz) [reply tensorboard/webapp/settings/polymer_interop_container.ts]: Makes sense.  (now I will try to remember to do this myself)

File: tensorboard/webapp/app_container.ng.html (hunk 6 above)
Comment (psybuzz) [reply tensorboard/webapp/app_container.ng.html]: I'm fine with landing as-is for now, since there are only a handful of elements in this template.

I still prefer fewer points of customization, since adding support for more (when there is a use case) is always in our control.

If this template had 50 elements that customized some TB feature, we must remember to maintain configurations of all variants.  That requires relying on memory, rather than process.
